Streamline HistoryBuf interface

sirhcel opened the pull request:
An addendum to #610 for streamlining the interface of `HistoryBuf` even further. I'm opening this PR with two proposals for starting a discussion.

sirhcel committed:
Remove implementing Deref<Target = [T]> for HistoryBuf

There is already `as_slice()` which gives access to the underlying
backing array where the elements are not in the order of writing.

With automatic dereferencing, functions like `iter()` or `windows()` are
automatically available for `HistoryBuf` and are returning items in an
order which does not reflect the write order and will likely surprise
users which did not carefully read all the documentation on for
`HistoryBuf` like me.

sirhcel committed:
Replace HistoryBuf::as_slice() to as_unordered_slice()

Some method names already contain statements on ordering like
`oldest_ordered()`. In this context a method returning an unordered
slice should state this clearly.

sirhcel committed:
Replace HistoryBuf::write with push

HistoryBuf already uses extend in method names for ingesting multiple
items at once and other heapless containers use push for methods
ingesting a single item. Let's be more consistent with this naming
scheme by switching from write to push here as well.

sgued commented:
LGTM, but is a breaking change like https://github.com/rust-embedded/heapless/pull/610. Added to https://github.com/rust-embedded/heapless/issues/613

max-heller commented on src/history_buf.rs:
```diff
@@ -572,18 +591,10 @@ impl<T, S: HistoryBufStorage<T> + ?Sized> Drop for HistoryBufInner<T, S> {
-        self
+        self.as_unordered_slice()
     }
 }
```
`AsRef<[T]>` where the `[T]` is unordered also feels like a footgun, is this something that could be removed or is there an important reason to keep it?

sirhcel replied:
Yes, I see that this is not what I would expect from `AsRef<[T]>`. To my understanding, the result was already unordered before and I did not think about that when just getting the return type into the right shape.

Well, I am in favor of removing this trait implementation as I don't see a way in the sense of the low/neglect-able runtime costs `AsRef` suggests.

sirhcel replied:
While at it, i removed implementing `AsRef<[T]>` with [b82c603](https://github.com/rust-embedded/heapless/pull/611/commits/b82c60372a256e29b8f618a7a5589c29008b9fe5).

max-heller commented on src/history_buf.rs:
```diff
@@ -592,7 +603,7 @@ where
     T: fmt::Debug,
 {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        <[T] as fmt::Debug>::fmt(self, f)
+        <[T] as fmt::Debug>::fmt(self.as_unordered_slice(), f)
     }
 }
```
Could we make `Debug` format the elements in order? Printing `HistoryBuf` as the underlying slice without including the write head is confusing

sirhcel replied:
This sounds pretty reasonable to me. Added this with [9b2d27a](https://github.com/rust-embedded/heapless/pull/611/commits/9b2d27a35536549911904d559a58951abeec03ce).

sirhcel committed:
Add test for an ordered debug representation for HistoryBuf

As @max-heller noted in
https://github.com/rust-embedded/heapless/pull/611/changes#r3507840148,
there is little point in having an unordered debug representation
without showing the write head as well.

sirhcel committed:
Switch to an ordered debug representation of HistoryBuf

There is already support for generating a default list representation
for an iterator. Let's just it with our `oldest_ordered()` iterator.

sirhcel committed:
Remove implementing AsRef<[T]> for HistoryBuf
